### PR TITLE
[front] feat: support pasting URL and UID in search bar

### PR DESCRIPTION
**Pull request by yuyujijin**
**related issues** #1268

---

### Description

Added support for search by directly typing video URL or video UID. 
Added test to verify it works as intended.

### Checklist

- [X] I added the related issue(s) id in the related issues section (if any)
  - if not, delete the related issues section
- [x] I described my changes and my decisions in the PR description
- [X] I read the development guidelines of the [CONTRIBUTING.md][development-guidelines]
- [X] The tests pass and have been updated if relevant
- [X] The code quality check pass  (there's actually none for front-end)

❤️ Thank you for your contribution!

[development-guidelines]: https://github.com/tournesol-app/tournesol/blob/main/CONTRIBUTING.md#development-guidelines

### CODE REVIEW

File: frontend/src/features/frame/components/topbar/Search.tsx [hunk 1]
```diff
@@ -55,7 +55,13 @@ const Search = () => {
     searchParams.delete('search');
     searchParams.append('search', search);
     searchParams.delete('offset');
-    history.push('/recommendations?' + searchParams.toString());
+
+    // Parse the searchParams to see if it matches a youtube video
+    // And if so stripes it UID and automatically push to the specific Tournesol page
+    const videoId = extractVideoId(search);
```
Comment (amatissart) [inline frontend/src/features/frame/components/topbar/Search.tsx:61]: One issue when using `extractVideoId()` on user input is that any 11-letters word will be considered as a YouTube video id. For example searching for "algorithmes" will redirect to `/entities/yt:algorithms` which is unexpected :worried:.

Maybe we should add an option to `extractVideoId()` to require the `yt:` prefix in this case. As video ids (without prefix `yt:`) are already present in the search index, the video would still be present in the search results, so I think that would be an acceptable trade-off. 

Note that the implementation of `extractVideoId()` has been slightly modified in #1847 and should be easier to maintain. Make sure to rebase on the latest "main" if you want to modify it. And rebasing would also fix the flaky test that caused the CI pipeline to fail (fixed in #1840)

File: tests/cypress/e2e/frontend/search.cy.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            cy.intercept('GET', `/polls/videos/entities/yt:${videoUID}/**`).as('videopoll')
+
+            // Visit main page (should work on every page since search bar is the same)
+            cy.visit('/');
+
+            // Input the video URL
+            cy.get('input[id="searchInput"]').click().type(`${videoURL}{enter}`);
+
+            // Check if AJAX returns 200
+            cy.wait(`@videopoll`).its('response.statusCode').should('eq', 200)
```
Comment (amatissart) [inline tests/cypress/e2e/frontend/search.cy.ts:17]: that's very nice to check the correct request has been triggered, but I would also expect to check that the user has been redirected to the expected page. That's more directly related to the feature we want to test: 

```suggestion
            // Input the video URL
            cy.get('input[id="searchInput"]').click().type(`${videoURL}{enter}`);
            
            // Redirection to video page
            cy.url().should('include', `/entities/yt:${videoUID}`)

            // Check if AJAX returns 200
            cy.wait(`@videopoll`).its('response.statusCode').should('eq', 200)
```

**Review by amatissart — COMMENTED** [review]
Thanks for this PR!  
The implementation behaves as expected, except in one edge case. See the comments.

**Comment by yuyujijin** [comment]
Hi @amatissart,
the idea of adding an option to require the `yt:` prefix seems to be a good compromise !

If I understood correctly, does the option should be **a function args** (that could be false by default) that forces the user to put `yt:<id>` when searching for videos by id in order to fix the 11 character problem ?

```
export function extractVideoId(idOrUrl: string, needYtPrefix = false)
```

**Comment by amatissart** [comment]
> If I understood correctly, the option should be **a function args** (that could be false by default) that forces the user to put `yt:<id>` when searching for videos by id in order to fix the 11 character problem ?

Yes that sounds good to me :+1:

**Comment by yuyujijin** [comment]
> > If I understood correctly, the option should be **a function args** (that could be false by default) that forces the user to put `yt:<id>` when searching for videos by id in order to fix the 11 character problem ?
> 
> Yes that sounds good to me 👍

So just to be sure : you have to put `yt:` when searching by UID or when wanting to be redirected (the feature I'm implementing) ? I think I need to be sure what is the desired behaviour in order not to break anything haha

**Comment by yuyujijin** [comment]
> Here is what I would expect, if that's looks good to you.
> 
> Query type	Example	Expected behavior of the search bar
> Video URL	`https://www.youtube.com/watch?v=xbBsSEProGY`	Redirected to entity page on Tournesol.
> Tournesol UID	`yt:xbBsSEProGY`	Redirected to entity page on Tournesol.
> Youtube video ID	`xbBsSEProGY`	Unchanged.
> Redirected to search results which contain the video.
> 11-letter word	`algorithmes`	Unchanged.
> Redirected to search results which contain videos related to the query.

perfectly clear ! working on it :))

**Comment by GresilleSiffle** [comment]
Thank you @yuyujijin for the last comments.

We will probably review and merge your work tomorrow :+1: 

If there is no objection from you, we will probably remove the comments that describe self explanatory code.

**Pull request merged**